### ISSUE TITLE
R.validate() should use the static functions (for images)

### ISSUE THREAD

**Issue by Tvede-dk** [mac-cain13/R.swift#161]
The validate function does not use the R.swift functions to retrive the images. 
Eg, if I have an image named "test", then the current validate function contains

`if UIImage(named: "test") == nil { throw ... }`

But should (propperly) be 

`if R.image.test() == nil { throw ....`

The reason behind:
the validate method is meant for debug / test, and in a test case, one would like to use the static functions, in order to increase coverage (and of cause to test it).

**Comment by mac-cain13** [comment]
Thanks for using R.swift and opening an issue for this!

Using `UIImage(named:)` and not `R.image.*` is easier and more robust to implement, so that is why I used that. I guess it could also be done with `R.image.*`, but that code coverage doesn't feel like a really good argument to me. Personally I don't think you should try to test generated code, just like you don't test a library you use in your project.

Do you feel that using `R.image.*` is significantly better than the current method?

**Comment by Tvede-dk** [comment]
> Thanks for using R.swift and opening an issue for this!

The honour is on my side, R.swift is super pleasent and a missing thing in the xcode build system. 

Ah I was considering that the reason behinde using `UIImage(named:)` robustness.
However, I do belive that testing generated code is a very good idea (#1 reason, is that "I own" the genreated code, and even though I trust libraries and generated code, the possiblility of a bug is still presented). 
Secondly I belive that one should test what one uses, and in this case, I am using the `R.swift.Image.*`, meaning that even if  `UIImage(named:)` worked , that does not mean that the `R.swift.Image.*` would (say some corner case, or the iOS / OSX sdk changed behavior ) 

I would suggest that either the `UIImage(named:)` changes to the constructor used in the `R.swift.Image.*` functions , or simply using the `R.swift.Image.*` 
(Or put more simple, to test what we are using rather than if it works, because it does work :D  )

**Comment by mac-cain13** [comment]
Thanks for the reply!

After thinking a bit more about it, I think it's a nice to have. I don't really see it as a major improvement, but it would be nice to use the `R.image.*` methods in the `validate` methods instead of Apples constructor. If you or someone else wants to pick it up and create a PR, I'm open to it!

**Comment by Tvede-dk** [comment]
I will create a PR when i have verified my change is working as expected :)

**Comment by tomlokhorst** [comment]
There is a (small) downside to using the generated code to validate conditions.

For example, if this is the code in `validate`:

`if R.image.test() == nil { throw ....`

And a future version of R.swift changes `R.image.test()` to return a `NSURL?`, this validate function would suddenly mark missing images as being ok, if there's only a NSURL.

Now, I don't think is likely that this particular example of `R.image.*` would suddenly change, as that would break a bunch of code.

But in general, using the same generated code to validate something the generated code depends on, is less robust. Compared to using two separate implementations.

Now, I'm all for tests and all. But in my opition testing should be part of a separate test project. It shouldn't be the goal of the runtime validate function that validates preconditions.

**Comment by Tvede-dk** [comment]
>.... But in general, using the same generated code to validate something the generated code depends on, is less robust. Compared to using two separate implementations.

Well this (small) downside, is in some senes also currently presented, and the more correct solution would be to proppergate the function object itself to the validate method, such that it would not depend on the hardcoded value, but that would propperly be a major change.

> Now, I'm all for tests and all. But in my opition testing should be part of a separate test project. It shouldn't be the goal of the runtime validate function that validates preconditions.

But is that not the point ? to validate at runtime, that retriving the object works ? Say for example that in the future you would have a loading / caching mechanism; then the validate method should use that, as the target application uses that. ( I belive that the validate is a "can i retrive all data at runtime", and that retrive is though the R.swift and not though the general loading mechanisms). 

But I would agree that either creating a seperate test project is a good idea, and then validate all the generated code there.  :)

**Comment by tomlokhorst** [comment]
The current behaviour of `validate` is to validate that the runtime environment matches the environment that existed when the `R.generated.swift` file was generated. 

There are two big examples of when `validate` might fail:

1. The `R.generated.swift` file is out of date, e.g. when R.swift is not part of the build, but is run manually.
2. Images are loaded as On-Demand Resources, and are not yet downloaded.

Note that in neither of these two cases, there is something wrong with the generated code. It's just that the environments don't match.

If you're using generated code in `validate`, you're conflating two separate concerns; Validating the environment, and testing the generated code.

I can see a use for testing generated code, but that would be as a separate function, e.g.:

```swift
R.validate() // Validate that the current runtime environment matches the expected environment
R.testNonNil() // Given a correct environment, test if all generated code returns non-nil values
```

**Comment by mac-cain13** [comment]
@tomlokhorst Good and valid points. `R.testNonNil()` is also an interesting proposal.

I have another major concern regarding the proposed change of @Tvede-dk that has to be fixed before we could merge in the change:

Images that are ignored by R.swift can't be in the validate method anymore, the test project contains two images named "Second" so R.swift emits a warning and doesn't generate `R.image.second()`. The image is however used in the storyboard, the generated code in the PR doesn't compile in this case because it generates code that uses the non-existing `R.image.second()`.

This also will be the case with images with an invalid name, for example "@" (see #164). This currently results in a compile error anyway, but when that is fixed this change will introduce a new problem.

I'm afraid solving this problem will either involve adding a lot of code in R.swift or means just skipping these images making the validation less useful.

What do you thing @Tvede-dk?

**Comment by Tvede-dk** [comment]
I would agree to @tomlokhorst as well, that is indeed a good point. But in the case that the R.swift file is out of date, would'nt that be a very big issue itself ?  using a auto generated file, but without updating it, seems to me as plain wrong. This would propperly mean a compile time error, instead of a runtime error (say one does not use the validate function, or has it turned off or does not look into the console), then the user would know that the generated file is out of sync / invalid in the environment.

Now Im no expert , but why would an image be ignored by R.swift ? and in the case one is ignored, then it is not in the hands of R.swift, meaning that R.swift should not assume anything about it ? (like validating its exsistence). 

Bottom line though: 
- its propperly best to discard the PR, since solving the bugs / cases with accepting it, and or making changes to which images are ignored and what not, is definitely something that should be thoroughly thought through.

**Comment by mac-cain13** [comment]
Yes, I agree that not update the generated file is another class of problem altogether. If that is the case all kinds of error could occur, but still the generated code should compile and run without crashes. Only the validate will start complaining.

By ignoring an image I mean either;
- detecting a duplicate image name, warning about it and skipping generation of a method/var in the generated file
- detecting an image with a name that cannot be converted to Swift, warning about it and skipping generation of a method/var in the generated file

Images are never "just ignored" it's always because there is a problem.

I agree it's best to discard to PR for now. Many thanks for your time and involvement! Much appreciated! :)
